api/millenium_patron.py

```diff
@@ -136,24 +168,47 @@ def __init__(self, library, integration, analytics=None):
         self.auth_mode = auth_mode
 
         self.block_types = integration.setting(self.BLOCK_TYPES).value or None
+
+        # Grab the patron record restriction information.
+        _db = Session.object_session(library)
+        self.patron_field_restriction_field = ConfigurationSetting.for_library_and_externalintegration(
+            _db, self.PATRON_FIELD_RESTRICTION_FIELD, library, integration
+        ).value
+        restriction = ConfigurationSetting.for_library_and_externalintegration(
+            _db, self.PATRON_FIELD_RESTRICTION_STRING, library, integration
+        ).value
+        if restriction and restriction.startswith("^"):
+            # Interpret the value a regular expression
+            try:
+                restriction = re.compile(restriction)
+            except Exception, e:
+                self.log.error(
+                    "Could not interpret patron record restriction as a regular expression: %r", e
+                )
+        self.patron_field_restriction = restriction
         
     # Begin implementation of BasicAuthenticationProvider abstract
     # methods.
 
     def _request(self, path):
         """Make an HTTP request and parse the response."""
-    
+
     def remote_authenticate(self, username, password):
         """Does the Millenium Patron API approve of these credentials?
 
         :return: False if the credentials are invalid. If they are
         valid, a PatronData that serves only to indicate which
         authorization identifier the patron prefers.
         """
+        # Check that we match the patron record restriction
+        patrondata, match = self.patron_record_restriction_matches(username)
+        if not match:
+            return False
+
         if not self.collects_password:
             # We don't even look at the password. If the patron exists, they
             # are authenticated.
-            patrondata = self._remote_patron_lookup(username)
+            patrondata = patrondata if patrondata else self._remote_patron_lookup(username)
```

Comment:
Won't patrondata always be defined here? If for some reason it wasn't, it probably wouldn't help to do the remote patron lookup again.